### PR TITLE
I3config

### PR DESCRIPTION
**Pull request by Chr1stoph**
https://github.com/i3/i3/issues/2141
Changes are related to master.
Generation of the i3-config file is not stopped in case a key symbol could not be resolved. The relevant line is copied with the unresolved number but appended with a comment. When i3 checks the bindings i3-nagbar is triggered in case of an unresolved key mapping.

### CODE REVIEW

File: i3-config-wizard/main.c [hunk 1]
```diff
@@ -264,29 +264,40 @@ static char *next_state(const cmdp_token *token) {
             const xkb_keysym_t *syms;
             int num = xkb_keymap_key_get_syms_by_level(xkb_keymap, keycode, 0, 0, &syms);
             if (num == 0)
-                errx(1, "xkb_keymap_key_get_syms_by_level returned no symbols for keycode %d", keycode);
-            if (!keysym_used_on_other_key(syms[0], keycode))
-                level = 0;
+                fprintf(stderr, "xkb_keymap_key_get_syms_by_level returned no symbols for keycode %d\n", keycode);
+            if (num != 0)
```
Comment (Airblader) [inline i3-config-wizard/main.c]: This should just be an `else` to the above condition. Also, please use the correct formatting (use `clang-format` on it). The Travis build will fail with incorrect formatting.

**Comment by Chr1stoph** [comment]
Ok, I think I have to learn https://travis-ci.org/ about the pull request work flow.

**Comment by Airblader** [comment]
No, you don't really need to understand Travis (just to save you that work). :) Just use `clang-format` to format the code (or do it by hand, it's really only about placement of braces as far as I can tell). To make changes to the pull request, simply amend to the commit and force-push the branch.

**Comment by Chr1stoph** [comment]
I have found `clang-format` on my system. Do you have a recommended configuration which I can apply using the -style option? At least IndentWidth and ColumnLimit do not match to the defaults I have here. I think so because `clang format` with my local defaults has more remarks on different parts of the code.

**Comment by Airblader** [comment]
@Chr1stoph You'll need to use clang-format from clang35, specifically. Newer versions won't work right, unfortunately. Having that, just `cd` into the repository and use

````
clang-format-3.5 -i $(find . -name "*.[ch]" | tr '\n' ' ')
```

We have our own style settings as part of the repository, so if executed from the correct path, those will be used.

**Comment by Chr1stoph** [comment]
@Airblader, please excuse me that it took so long to apply the change. I hope Travis is happy now.

**Comment by Airblader** [comment]
@stapelberg Ping

File: i3-config-wizard/main.c [hunk 2]
```diff
@@ -263,30 +263,46 @@ static char *next_state(const cmdp_token *token) {
...
         if (num > 1)
-            printf("xkb_keymap_key_get_syms_by_level (keycode = %d) returned %d symbolsinstead of 1, using only the first one.\n", keycode, num);
-
+            fprintf(stderr, "xkb_keymap_key_get_syms_by_level (keycode = %d) returned %d symbolsinstead "
+                            "of 1, using only the first one.\n",
+                    keycode, num);
         char str[4096];
-        if (xkb_keysym_get_name(syms[0], str, sizeof(str)) == -1)
-            errx(EXIT_FAILURE, "xkb_keysym_get_name(%u) failed", syms[0]);
+        char warn[90];
```
Comment (stapelberg) [inline i3-config-wizard/main.c]: Please use `char *warn = "";` here and `warn = " # ERROR:";` below instead of `strcpy` and magic numbers :)

**Comment by stapelberg** [comment]
Please also squash your commits and give them a proper title/description.

File: i3-config-wizard/main.c (hunk 2 above)
Comment (Chr1stoph) [inline i3-config-wizard/main.c]: Ok, I have had to apply char *warn = ""; above the if statement to have it defined for all cases of num. This is much cleaner than my approach. An alternate solution would be a const char *warn = " # ERROR: ..."; and sasprintf(&res, ..., (num != 0 ? "" : warn));. This style is really elegant. Now I will upload the file main.c with your improvement.

**Comment by Chr1stoph** [comment]
@stapelberg I have no idea about the travis result. If I am not wrong the details show the travis output but no claims with respect to the file main.c. Here clang-format35 -style=file on main.c seems to be ok. What shall I do?

**Comment by stapelberg** [comment]
> @stapelberg I have no idea about the travis result. If I am not wrong the details show the travis output but no claims with respect to the file main.c. Here clang-format35 -style=file on main.c seems to be ok. What shall I do?

The issue is not formatting, but the fact that travis can’t find the base image (and new base images cannot be pushed from pull requests). We’ve fixed this issue a while ago in commit 42f5a6ce479968a8f95dd5a827524865094d6a5c.

However, your pull request by now is so old that it doesn’t include that commit :).

Because of this, and because it’s generally good practice (to avoid merge conflicts), the right solution for this problem is to rebase your work onto the current “next” branch. See https://git-scm.com/book/en/v2/Git-Branching-Rebasing for more details.

**Comment by Chr1stoph** [comment]
On Wed, Jun 22, 2016 at 12:44:47AM -0700, Michael Stapelberg wrote:
> > @stapelberg I have no idea about the travis result. If I am not wrong the details show the travis output but no claims with respect to the file main.c. Here clang-format35 -style=file on main.c seems to be ok. What shall I do?
> 
> The issue is not formatting, but the fact that travis can’t find the base image (and new base images cannot be pushed from pull requests). We’ve fixed this issue a while ago in commit 42f5a6ce479968a8f95dd5a827524865094d6a5c.
> 
> However, your pull request by now is so old that it doesn’t include that commit :).
> 
> Because of this, and because it’s generally good practice (to avoid merge conflicts), the right solution for this problem is to rebase your work onto the current “next” branch. See https://git-scm.com/book/en/v2/Git-Branching-Rebasing for more details.

Ok, https://guides.github.com/introduction/flow/ and others stated to
merge from master. I am not sure if it is possible to rebase
to the "next" branch using the browser. If not I will have to
try the quite new git command line tools on FreeBSD.

Or is it possible to delete the branch and create
a correct one?

**Comment by Airblader** [comment]
@Chr1stoph All you have to do is

````
# unless already done, add the upstream i3 remote to fetch its current state
git remote add vanilla git@github.com:i3/i3.git
git fetch vanilla

# now rebase next onto your branch
# NOTE: this might cause conflicts which you need to resolve (see instructions by git)
git rebase vanilla/next

# force-push the changes to update the pull request
git push --force
```

**Comment by Chr1stoph** [comment]
On Wed, Jun 22, 2016 at 10:52:44AM -0700, Ingo B�rk wrote:
> @Chr1stoph All you have to do is
> 
> ````
> # unless already done, add the upstream i3 remote to fetch its current state
> git remote add vanilla git@github.com:i3/i3.git
> git fetch vanilla
> 
> # now rebase next onto your branch
> # NOTE: this might cause conflicts which you need to resolve (see instructions by git)
> git rebase vanilla/next
> 
> # force-push the changes to update the pull request
> git push --force
> ```
@Airblader I will try 
http://www.freshports.org/devel/git-lite/
I would be happy not to need the nice web interface anymore.

**Comment by Chr1stoph** [comment]
On Wed, Jun 22, 2016 at 08:51:01PM +0200, Christoph Brinkhaus wrote:
> On Wed, Jun 22, 2016 at 10:52:44AM -0700, Ingo B�rk wrote:
> > @Chr1stoph All you have to do is
> > 
> > ````
> > # unless already done, add the upstream i3 remote to fetch its current state
> > git remote add vanilla git@github.com:i3/i3.git
> > git fetch vanilla
> > 
> > # now rebase next onto your branch
> > # NOTE: this might cause conflicts which you need to resolve (see instructions by git)
> > git rebase vanilla/next
> > 
> > # force-push the changes to update the pull request
> > git push --force
> > ```
> @Airblader I will try 
> http://www.freshports.org/devel/git-lite/
> I would be happy not to need the nice web interface anymore.

Currently I hang here:
$ git config --list
user.name=Chr1stoph
user.email=c.brinkhaus@t-online.de
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
remote.vanilla.url=git@github.com:i3/i3.git
remote.vanilla.fetch=+refs/heads/*:refs/remotes/vanilla/*

$ git fetch vanilla
Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

I will read https://git-scm.com/.
Is think no need to link somehow to
https://github.com/Chr1stoph/i3/tree/i3config ?

**Comment by Airblader** [comment]
@Chr1stoph Using `git@github.com:i3/i3.git` will only work if you use an SSH key with git that is associated to your Github account. If you don't, don't worry, just do

````
git remote remove vanilla
git remote add vanilla https://github.com/i3/i3.git

# and continue as above
```

**Comment by Chr1stoph** [comment]
On Wed, Jun 22, 2016 at 01:01:23PM -0700, Ingo B�rk wrote:
> @Chr1stoph Using `git@github.com:i3/i3.git` will only work if you use an SSH key with git that is associated to your Github account. If you don't, don't worry, just do
> 
> ````
> git remote remove vanilla
> git remote add vanilla https://github.com/i3/i3.git
@Airblader This worked well. Now I have a directory .git with the
entries FETCH_HEAD, HEAD, branches, config, description,
hooks, info, logs, objects, refs

There is no code, there are references or so to many i3 versions.
> 
> # and continue as above
> ```
Now the next output is $ git rebase vanilla/next
fatal: Needed a single revision

Currently I do not understand where the reference to the
pull is which should be on
https://github.com/Chr1stoph/i3/tree/i3config

Without this reference I have no idea how git should find
out what to rebase. My guess is that there must be either
a pointer to the address above. Or I have to synchronize
the folder where I have launched the git commands with
the address above, may be by git-clone or git-fetch.

I think I miss something very basic. The kind of mistake
one does only once...

**Comment by Airblader** [comment]
Sorry, I assume you were working in a local checkout of your own repository. If that's not the case, let's just start "fresh".


````
# Or change to anything else, since this one will be wiped upon reboot.
cd /tmp

# let's clone your own repository and check out the feature branch
git clone https://github.com/Chr1stoph/i3.git
git fetch --all
git checkout i3config

# now we do what we wanted to do above
git remote add vanilla https://github.com/i3/i3.git
git fetch vanilla
git rebase vanilla/next
git push --force
```

**Comment by Chr1stoph** [comment]
On Thu, Jun 23, 2016 at 12:07:30PM -0700, Ingo B�rk wrote:
> Sorry, I assume you were working in a local checkout of your own repository. If that's not the case, let's just start "fresh".
> 
> 
> ````
> # Or change to anything else, since this one will be wiped upon reboot.
> cd /tmp
> 
> # let's clone your own repository and check out the feature branch
> git clone https://github.com/Chr1stoph/i3.git
@Airblader to took me lots of trials to find out that
I need to do a simple cd i3 here. The good thing is that I
learned somehow and start to use git for my local things;-).
> git fetch --all
> git checkout i3config
> 
> # now we do what we wanted to do above
> git remote add vanilla https://github.com/i3/i3.git
> git fetch vanilla
> git rebase vanilla/next
This gives some errors as expected. But I am not sure how to
fix everything. The missing patch appears in .git/rebase-apply/patch.
The first example is the user guide where the diff is

diff --git a/docs/userguide b/docs/userguide
index
13dae4fe087f5798ae05cc49cf4ed6fce60abba1..b077eb66d42e49cc9a70361c615be4a9691e9b10
100644
--- a/docs/userguide
+++ b/docs/userguide
@@ -2134,10 +2134,10 @@ and the following placeholders which will be
replaced:
 
 +%title+::
 The X11 window title (_NET_WM_NAME or WM_NAME as fallback).
 -+%class+:
 ++%class+::
 The X11 window class (second part of WM_CLASS). This corresponds to
 the 
 +class+ criterion, see <<command_criteria>>.
 -+%instance+:
 ++%instance+::
 The X11 window instance (first part of WM_CLASS). This corresponds to
 the 
 +instance+ criterion, see <<command_criteria>>.

 The only line in i3/docs/userguide where the word "placeholder"
 appears is in line 2217. This is likely difficult to patch because of
 the huge offset. I do not understand where the diff comes from.

 I think I can download the userguide out of the next branch
 and try to make the version of my pullrequest look less different.
 Will that cause the patch file be different?
 What happens if i simply take the next version and replace
 the version of the pullrequest with that one?

 Later there are also some patches of *.c files which I need to
 handle as well. I have not yet read the section about merge in
 https://git-scm.com/book. May be this gives answers, too.

**Comment by Airblader** [comment]
> The missing patch appears in .git/rebase-apply/patch.

On a general note, you don't need to worry about the `.git` folder at all. It's just for git-internal things. :-)

When you executed the rebase command and run into trouble, you now need to use `git mergetool`. For that you should configure it with something decent (though I personally use `vimdiff` for the mergetool) and you should definitely read up on how to merge changes first. This is the part where I can't just write down commands – you now have to combine changes correctly. :-)

You can perhaps also try rebasing out in a tutorial like [this one](http://learngitbranching.js.org/) first (or how to use git in general). It's a pretty nice thing to do!

**Comment by Chr1stoph** [comment]
On Thu, Jun 30, 2016 at 02:07:15PM -0700, Ingo Bürk wrote:
> > The missing patch appears in .git/rebase-apply/patch.
@Airblader 
> On a general note, you don't need to worry about the `.git` folder at all. It's just for git-internal things. :-)
The message mentioned that file. It is good to know that I do not have
to care about its internals.
> 
> When you executed the rebase command and run into trouble, you now need to use `git mergetool`. For that you should configure it with something decent (though I personally use `vimdiff` for the mergetool) and you should definitely read up on how to merge changes first. This is the part where I can't just write down commands – you now have to combine changes correctly. :-)
To be able to use vimdiff is a good news.

> You can perhaps also try rebasing out in a tutorial like [this one](http://learngitbranching.js.org/) first (or how to use git in general). It's a pretty nice thing to do!
I will check the link. Thank you for your patiance.

**Comment by Chr1stoph** [comment]
On Fri, Jul 01, 2016 at 05:48:57PM +0200, Christoph Brinkhaus wrote:
> On Thu, Jun 30, 2016 at 02:07:15PM -0700, Ingo Bürk wrote:
> > > The missing patch appears in .git/rebase-apply/patch.
> @Airblader 
> > On a general note, you don't need to worry about the `.git` folder at all. It's just for git-internal things. :-)
> The message mentioned that file. It is good to know that I do not have
> to care about its internals.
> > 
> > When you executed the rebase command and run into trouble, you now need to use `git mergetool`. For that you should configure it with something decent (though I personally use `vimdiff` for the mergetool) and you should definitely read up on how to merge changes first. This is the part where I can't just write down commands – you now have to combine changes correctly. :-)
> To be able to use vimdiff is a good news.
> 
> > You can perhaps also try rebasing out in a tutorial like [this one](http://learngitbranching.js.org/) first (or how to use git in general). It's a pretty nice thing to do!
> I will check the link. Thank you for your patiance.
>
@Airblader First I did some tests with rebase on a local folder to get
used to the procedure. Strange has be that src/bindings.c as the only
file in the test repository rebased without interaction. But it is
clearly useful to get an idea how things are shown in vimdiff launched
by git mergetool.

Now everything might have been successfull. I have to check if the
changes are in, if it can be build and if everything finally works.

**Comment by Chr1stoph** [comment]
On Tue, Jul 05, 2016 at 08:22:28PM +0200, Christoph Brinkhaus wrote:
> On Fri, Jul 01, 2016 at 05:48:57PM +0200, Christoph Brinkhaus wrote:
> > On Thu, Jun 30, 2016 at 02:07:15PM -0700, Ingo Bürk wrote:
> > > > The missing patch appears in .git/rebase-apply/patch.
> > @Airblader 
> > > On a general note, you don't need to worry about the `.git` folder at all. It's just for git-internal things. :-)
> > The message mentioned that file. It is good to know that I do not have
> > to care about its internals.
> > > 
> > > When you executed the rebase command and run into trouble, you now need to use `git mergetool`. For that you should configure it with something decent (though I personally use `vimdiff` for the mergetool) and you should definitely read up on how to merge changes first. This is the part where I can't just write down commands – you now have to combine changes correctly. :-)
> > To be able to use vimdiff is a good news.
> > 
> > > You can perhaps also try rebasing out in a tutorial like [this one](http://learngitbranching.js.org/) first (or how to use git in general). It's a pretty nice thing to do!
> > I will check the link. Thank you for your patiance.
> >
> @Airblader First I did some tests with rebase on a local folder to get
> used to the procedure. Strange has be that src/bindings.c as the only
> file in the test repository rebased without interaction. But it is
> clearly useful to get an idea how things are shown in vimdiff launched
> by git mergetool.
> 
> Now everything might have been successfull. I have to check if the
> changes are in, if it can be build and if everything finally works. 

@Airblader Now there is the next problem. in i3config
src/config_parser.c includes xcb/xcb_xrm.h. There are also other lines
where xcb_xrm appears. In the original sorces of FreeBSD xcb_xrm does
not appear anywhere.

Of course it does not compile because of the missing header file.
But I guess src/config_parser.c is corrupted somehow.

When I check src/config_parser.c using the browser xcb_xrm only
appears in i3config. Not in next and not in master. Do you have any
idea what I could have done wrong? And how to fix it?

**Comment by Airblader** [comment]
> Now everything might have been successfull.

No, the PR is definitely broken right now. It contains 90 commits and over 100 changed files, which is definitely not correct.

> In the original sorces of FreeBSD xcb_xrm does
not appear anywhere.

What FreeBSD sources and why are you looking at anything from FreeBSD? There is only one source code against which you should provide a PR and that's the repository hosted here on Github. xcb_xrm being in the code is definitely correct.

Sorry, I'm really not sure what you're doing. You fork the Github repository, clone your own fork, make your changes against it on some branch, push that branch and create the PR. If things changed here, you add it as a separate remote, rebase the branch and resolve conflicts. That's really all that has to happen, but you must've done something else.

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 11:40:31AM -0700, Ingo B�rk wrote:
> > Now everything might have been successfull.
> 
> No, the PR is definitely broken right now. It contains 90 commits and over 100 changed files, which is definitely not correct.
> 
> > In the original sorces of FreeBSD xcb_xrm does
> not appear anywhere.
> 
> What FreeBSD sources and why are you looking at anything from FreeBSD? 
Because it runs on FreeBSD without special patches or so. Therefore I
think it should work again when I compile it under FreeBSD. Of course
the repository is github.
> There is only one source code against which you should provide a PR and that's the repository hosted here on Github. xcb_xrm being in the code is definitely correct.

Ok, xcb_xrm appears in
https://github.com/i3/i3/tree/next/src/config-parser.c
But I can not find the file xcb/xcb_xrm.h. 

I guess
https://github.com/Chr1stoph/i3/blob/next/src/config_parser.c
is from the time when I started the pull request or so.
> 
> Sorry, I'm really not sure what you're doing. You fork the Github repository, clone your own fork, make your changes against it on some branch, push that branch and create the PR. If things changed here, you add it as a separate remote, rebase the branch and resolve conflicts. That's really all that has to happen, but you must've done something else.
It is strange, especially since I changed just two files.

**Comment by Airblader** [comment]
> But I can not find the file xcb/xcb_xrm.h.

That's because it's not a file from i3, it's an external library that I wrote.

> It is strange, especially since I changed just two files.

It's the result of some incorrect rebase / merge. But I didn't sit next to you, so I can't really say what exactly went wrong. Maybe at some point it'd just be better to start with a new feature branch and redo the changes?

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 09:15:09PM +0200, Christoph Brinkhaus wrote:
> On Wed, Jul 06, 2016 at 11:40:31AM -0700, Ingo B�rk wrote:
> > > Now everything might have been successfull.
> > 
> > No, the PR is definitely broken right now. It contains 90 commits and over 100 changed files, which is definitely not correct.
> > 
> > > In the original sorces of FreeBSD xcb_xrm does
> > not appear anywhere.
> > 
> > What FreeBSD sources and why are you looking at anything from FreeBSD? 
> Because it runs on FreeBSD without special patches or so. Therefore I
> think it should work again when I compile it under FreeBSD. Of course
> the repository is github.
> > There is only one source code against which you should provide a PR and that's the repository hosted here on Github. xcb_xrm being in the code is definitely correct.
> 
> Ok, xcb_xrm appears in
> https://github.com/i3/i3/tree/next/src/config-parser.c
> But I can not find the file xcb/xcb_xrm.h. 
Now I found it here using a web search engine:
https://github.com/Airblader/xcb-util-xrm/tree/master/include
Does i3 requires xcb-util-xrm as a depenency?
Some of these utilities are installed here, too.
But it might be behind Debian.
> 
> I guess
> https://github.com/Chr1stoph/i3/blob/next/src/config_parser.c
> is from the time when I started the pull request or so.
> > 
> > Sorry, I'm really not sure what you're doing. You fork the Github repository, clone your own fork, make your changes against it on some branch, push that branch and create the PR. If things changed here, you add it as a separate remote, rebase the branch and resolve conflicts. That's really all that has to happen, but you must've done something else.
> It is strange, especially since I changed just two files.

**Comment by Airblader** [comment]
> Does i3 requires xcb-util-xrm as a depenency?

Yes. The current release, 4.12, does not yet have the dependency. The next release, 4.13, will have the dependency.

Debian already provides a package for xcb-util-xrm in Debian testing. I'm not aware of FreeBSD (or any BSD for that matter) having created a package for it yet.

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 12:17:32PM -0700, Ingo B�rk wrote:
> > But I can not find the file xcb/xcb_xrm.h.
> 
> That's because it's not a file from i3, it's an external library that I wrote.
> 
> > It is strange, especially since I changed just two files.
> 
> It's the result of some incorrect rebase / merge. But I didn't sit next to you, so I can't really say what exactly went wrong. Maybe at some point it'd just be better to start with a new feature branch and redo the changes?
I am happy that you suggest this. I think it should mean cleanup
i3config, clone next, apply the small changes.

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 12:25:13PM -0700, Ingo B�rk wrote:
> > Does i3 requires xcb-util-xrm as a depenency?
> 
> Yes. The current release, 4.12, does not yet have the dependency. The next release, 4.13, will have the dependency.
Currently there is
x11/xcb                   Tool for managing x11 cut-buffers
x11/xcb-proto             The X protocol C-language Binding (XCB) protocol
x11/xcb-util              Module with libxcb/libX11 extension/replacement libraries
x11/xcb-util-cursor       XCB cursor library
x11/xcb-util-image        Port of Xlib's XImage and XShmImage functions
x11/xcb-util-keysyms      Standard X key constants and conversion to/from keycodes
x11/xcb-util-renderutil   Convenience functions for the Render extension
x11/xcb-util-wm           Framework for window manager implementation

> 
> Debian already provides a package for xcb-util-xrm in Debian testing. I'm not aware of FreeBSD (or any BSD for that matter) having created a package for it yet.
May be xcb-util-xrm is more important than my lousy small patch.
I will contact the maintainer of x11/xcb-util
and ask for the opinion. i3 has more than one FreeBSD user.

**Comment by Airblader** [comment]
> May be xcb-util-xrm is more important than my lousy small patch.

I don't understand what you mean with that. The next i3 release will have this dependency, that's a patch that's already been merged. I have no idea what source of i3 you are looking at where xcb-util-xrm is not used. Whichever it is, it's not up to date. The *only* repository that matters for pull requests is *this* one.

> I will contact the maintainer of x11/xcb-util

What for?

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 12:53:28PM -0700, Ingo B�rk wrote:
> > May be xcb-util-xrm is more important than my lousy small patch.
> 
> I don't understand what you mean with that. The next i3 release will have this dependency, that's a patch that's already been merged. I have no idea what source of i3 you are looking at where xcb-util-xrm is not used. Whichever it is, it's not up to date. The *only* repository that matters for pull requests is *this* one.
> 
> > I will contact the maintainer of x11/xcb-util
> 
> What for?
If i3 will have a version 4.13 and requires x11/xcb-util-xrm which
does not existi, this is a problem for FreeBSD.
Regarding i3 the source is as usual. But the dependency must be
fullfilled. Please excuse me if I am wrong. From my understanding
xcb-util-xrm is not provided by i3 but from a different repository.

**Comment by Airblader** [comment]
No, you are correct. But they will notice this when trying to provide the i4-4.13 package because they won't be able to build. But yeah, please do let them know upfront, that certainly doesn't hurt.

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 01:08:55PM -0700, Ingo B�rk wrote:
> No, you are correct. But they will notice this when trying to provide the i4-4.13 package because they won't be able to build. But yeah, please do let them know upfront, that certainly doesn't hurt.
Ok, the maintainer of i3 already knows about that.
Currently I ask myself what I am doing here...

Christoph

**Comment by Chr1stoph** [comment]
On Wed, Jul 06, 2016 at 09:29:32PM +0200, Christoph Brinkhaus wrote:
> On Wed, Jul 06, 2016 at 12:17:32PM -0700, Ingo B�rk wrote:
> > > But I can not find the file xcb/xcb_xrm.h.
> > 
> > That's because it's not a file from i3, it's an external library that I wrote.
@Airblader The next brach of i3 works now including the xcb_xrm
library.
> > 
> > > It is strange, especially since I changed just two files.
> > 
> > It's the result of some incorrect rebase / merge. But I didn't sit next to you, so I can't really say what exactly went wrong. Maybe at some point it'd just be better to start with a new feature branch and redo the changes?
> I am happy that you suggest this. I think it should mean cleanup
> i3config, clone next, apply the small changes.
The complete repository I have forked seems to be difficult.
The "next" branch of the fork is completely different from
the "next" branch of i3. Is it possible to synchronize them?

If not I am not sure if it is the best to close the current PR,
destroy the fork and create a completely new one.

**Comment by Chr1stoph** [comment]
On Wed, Jul 20, 2016 at 07:55:22PM +0200, Christoph Brinkhaus wrote:
> On Wed, Jul 06, 2016 at 09:29:32PM +0200, Christoph Brinkhaus wrote:
> > On Wed, Jul 06, 2016 at 12:17:32PM -0700, Ingo B�rk wrote:
> > > > But I can not find the file xcb/xcb_xrm.h.
> > > 
> > > That's because it's not a file from i3, it's an external library that I wrote.
> @Airblader The next brach of i3 works now including the xcb_xrm
> library.
> > > 
> > > > It is strange, especially since I changed just two files.
> > > 
> > > It's the result of some incorrect rebase / merge. But I didn't sit next to you, so I can't really say what exactly went wrong. Maybe at some point it'd just be better to start with a new feature branch and redo the changes?
> > I am happy that you suggest this. I think it should mean cleanup
> > i3config, clone next, apply the small changes.
> The complete repository I have forked seems to be difficult.
> The "next" branch of the fork is completely different from
> the "next" branch of i3. Is it possible to synchronize them?
> 
> If not I am not sure if it is the best to close the current PR,
> destroy the fork and create a completely new one.
@Airblader I have found
https://gun.io/blog/how-to-github-fork-branch-and-pull-request/
which is from 2012 but hopefully still up to date.
I will try this procedure but copy next instead of master.

**Comment by Airblader** [comment]
I'm going to go ahead and close this PR now. @Chr1stoph When you have figured it out, please just open a new PR for the change there. This one has become too messy for my taste and progress is really slow, plus the incorrect rebase has put a lot of people on this issue.

**Comment by Chr1stoph** [comment]
On Sat, Aug 20, 2016 at 06:02:23AM -0700, Ingo B�rk wrote:
> I'm going to go ahead and close this PR now. @Chr1stoph When you have figured it out, please just open a new PR for the change there. This one has become too messy for my taste and progress is really slow, plus the incorrect rebase has put a lot of people on this issue.
@Airblader, this is fine with me. I hope I can clean up everything
sooner or later. Thank you again for the support regarding github.

**Pull request closed without merging**